File: go/euler07.go
// euler07 находит 10001-е простое число.
//
// Решение задачи Эйлера №7.
package main

import (
	"fmt"
	"time"
)

func isPrime(n int) bool {
	if n < 2 {
		return false
	}
	if n == 2 {
		return true
	}
	if n%2 == 0 {
		return false
	}
	for i := 3; i*i <= n; i += 2 {
		if n%i == 0 {
			return false
		}
	}
	return true
}

func euler07() {
	n := 10001

	startTime := time.Now()
	count := 0
	num := 2
	for {
		if isPrime(num) {
			count++
			if count == n {
				fmt.Println(num)
				break
			}
		}
		num++
	}
	fmt.Println(time.Since(startTime))

	startTime = time.Now()
	count = 0
	num = 2
	for {
		if isPrime(num) {
			count++
			if count == n {
				fmt.Println(num)
				break
			}
		}
		num++
	}
	fmt.Println(time.Since(startTime))
}

func main() {
	euler07()
}

File: python/euler07.py
# Какое число является 10001-м простым числом?

from datetime import datetime
from itertools import count, islice
from typing import Generator, Optional


def primes(dimension: Optional[int] = None) -> Generator[int, None, None]:
    """
    Итерируемая функция генерации простых чисел.

    Args:
        dimension: Количество цифр в числах (None для всех простых чисел).

    Yields:
        Очередное простое число.
    """
    start = 2 if dimension is None else 10 ** (dimension - 1) + 1
    ints = count(start)
    while True:
        p = next(ints)
        yield p
        ints = filter(p.__rmod__, ints)


def euler07() -> None:
    """
    Решение задачи Эйлера №7.

    Какое число является 10001-м простым числом?
    """
    n = 10_001

    # 1 Вариант
    start_time = datetime.now()
    number = primes()
    for i in range(n):
        prime = next(number)
    print(prime)
    print(datetime.now() - start_time)

    # 2 Вариант
    start_time = datetime.now()
    print(list(islice(primes(), n))[-1])
    print(datetime.now() - start_time)


if __name__ == "__main__":
    euler07()
